pkg/api/validation/validation.go

```diff
@@ -1937,6 +1970,56 @@ func ValidateVolumeMounts(mounts []api.VolumeMount, volumes sets.String, contain
 	return allErrs
 }
 
+func ValidateVolumeDevices(devices []api.VolumeDevice, volumes []api.Volume, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	devicepath := sets.NewString()
+	devicename := sets.NewString()
+
+	if devices != nil && !utilfeature.DefaultFeatureGate.Enabled(features.BlockVolumeSupport) {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("volumeDevices"), "Container volumeDevices is disabled by feature-gate"))
+		return allErrs
+	}
+	if devices != nil {
+		for i, dev := range devices {
+			idxPath := fldPath.Index(i)
+			devName := dev.Name
+			devPath := dev.DevicePath
+			volSource, didMatch := getVolumeSourceFromName(devName, volumes)
+			if len(devName) == 0 {
+				allErrs = append(allErrs, field.Required(idxPath.Child("name"), ""))
+			}
+			if devicename.Has(devName) {
+				allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), devName, "must be unique"))
+			}
+			// Must be PersistentVolumeClaim volume source
+			if didMatch && volSource.PersistentVolumeClaim == nil {
+				allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), devName, "volumeDevice does have a matching volume name however, volumeDevices can only use volume source type of PersistentVolumeClaim (pvc)"))
```

Comment:
The msg could be simplifed to "can only use volume source type of PersistentVolumeClaim for block mode"